### DISCUSSION TITLE
Slice JAX ShapedArray using indices

### DISCUSSION THREAD

**Discussion by snat1505027** [jax-ml/jax#12435] [Q&A] [answered]
Hi, I have a JAX ShapedArray, `X`, of shape `(1, 1080, 768) -> (batch_size, seq_len, hidden_dim) `and another sorted ShapedArray,  `IDX`, of size `(1, 270)` where the values in the `IDX` represent the indices of `seq_len` in `X`. I want to create a ragged tensor, `Y`, where the values in the `X` will be replaced by `MASK` if that index appears in `IDX`.

For example, X = [[[ 9, 10,  3],
                                [ 7, 10,  0]]]
                      IDX = [[1]]
                      
                      then Y = [[[ 9, 10,  3],[MASK]]]
                      
Is there any efficient way to do it?

**Comment by jakevdp** [comment, marked as answer]
Can you say more about why you're working with `ShapedArray` objects? They are essentially just abstract descriptors that contain a shape and a dtype. They don't contain any data, so there's not really any way to "slice" them, or to reason about the values in them. Perhaps you could edit your question with a minimal executible example of the kind of thing you want to do?

**Reply by snat1505027** [reply]
So I am working with a model that has been built using tensorflow and I want to add this intermediate step. So when I checked those arrays (X and IDX) I found their type as ShapedArray. Let me know if it answers your question.

**Reply by jakevdp** [reply]
It sounds like you've printed the values within a function during some stage of abstract evaluation. You cannot operate on ShapedArray objects directly (read [How To Think In JAX](https://jax.readthedocs.io/en/latest/notebooks/thinking_in_jax.html) for some context on tracing of functions). Without more information, preferably in the form of a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example), I don't think we'll be able to answer your question.

**Reply by snat1505027** [reply]
So to clarify, I was able to do concatenation using `jnp.concatenate` among the ShapedArrays with `pdb.set_trace()`. I am confused what you meant by `cannot operate on ShapedArray objects directly`.  

Also the type of those arrays were taken using `pdb.set_trace()` and they were being used in a forward propagation of the model training.

**Reply by jakevdp** [reply]
What I mean is that if you have an object of type `jax.core.ShapedArray`, it is just an abstract value that describes the shape and dtype of an array. It is not an actual array containing data, and so you cannot perform array operations on it. Because you haven't included any code, I'm just going off your verbal description, in which you said you had a ShapedArray object, but if you were able to pass it to `jnp.concatenate`, then I am probably not understanding your description correctly.

Sorry if that's unhelpful, but I'm having trouble understanding exactly what is happening in your code. A minimal reproducible example would help me give you a better answer.

**Reply by snat1505027** [reply]
Sorry about the confusion. So the current state of my code is:

```
imgs_enc = self.vision_encoder(batch['images'].reshape(batch_size * images_per_batch, seq_size, img_dim))['seq_attnpool'] # (7, seq_size // 4, 768)
imgs_enc = imgs_enc.reshape(batch_size, images_per_batch, seq_size // 4, self.hidden_size) # (1, 7, seq_size // 4, 768)
num_ans_per = 1
imgs_enc = jnp.concatenate([
            jnp.zeros([batch_size, 1, seq_size // 4, self.hidden_size], dtype=imgs_enc.dtype),
            imgs_enc,], 1) # (1, 8, seq_size // 4, 768)
        
images_per_batch += 1
imgs_enc = imgs_enc.reshape(batch_size, images_per_batch * seq_size // 4, self.hidden_size).repeat(num_ans_per, axis=0) # (1, 1080, 768)

num_image_spans = images_per_batch* seq_size // 4
num_image_patches_trg = int(num_image_spans * 0.25)
image_patch_trg_idx = jnp.asarray(uniform_random_select(n=num_image_spans, num_samples=num_image_patches_trg, sort_idx=True)[None]) 
```
where `uniform_random_select()` is

```
def uniform_random_select(n, num_samples, sort_idx=True):
    """
    Randomly choose "num_samples" from N
    :param n:
    :param num_samples:
    :param sort_idx: Whether to sort the resulting index
    :return:
    """
    if isinstance(num_samples, int) and isinstance(n, int):
        assert num_samples <= n
    logits = tf.random.uniform([n])
    idx = tf.argsort(logits)[:num_samples]
    if sort_idx:
        idx = tf.sort(idx)
    idx = tf.cast(idx, dtype=tf.int32)
    return idx
```

Here `X` is the `imgs_enc` and `IDX` is the `image_patch_trg_idx` and I want to put `MASK` token in the indices mentioned in `image_patch_trg_idx` and keep the rest of the `imgs_enc` as it is. Let me know if this helps.

**Reply by jakevdp** [reply]
Thanks - it looks like you're using normal JAX arrays, so everything I said about ShapedArray is not applicable (the benefit of pasting code rather than describing it 😁 ). I don't see `MASK` anywhere in your code, though – can you include the snippet that defines it, and give an example of what you want to do with it? Often it helps to simplify the code to the essential pieces to create a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example)  – something I could paste into an editor and run (i.e. include all relevant imports, function definitions, and array definitions). The less I have to guess about your code and intent, the better I'll be able to answer your question. To be quite honest, I still don't have a very good idea of what you're trying to do.

**Reply by snat1505027** [reply]
```
import dataclasses

import jax
import jax.numpy as jnp
import numpy as np
import flax.linen as nn
from jax import lax
from typing import Any, Dict, Union, Optional
import math
from jax import jit
from copy import deepcopy
import clu.parameter_overview
from dataclasses import dataclass
from functools import partial
import tensorflow as tf
from jax import random

MASK = 3

batch_size = 1
vis_seq_len = 1080
hidden_dim = 768
mask_rate = 0.25

def uniform_random_select(n, num_samples, sort_idx=True):
    """
    Randomly choose "num_samples" from N
    :param n:
    :param num_samples:
    :param sort_idx: Whether to sort the resulting index
    :return:
    """
    if isinstance(num_samples, int) and isinstance(n, int):
        assert num_samples <= n
    logits = tf.random.uniform([n])
    idx = tf.argsort(logits)[:num_samples]
    if sort_idx:
        idx = tf.sort(idx)
    idx = tf.cast(idx, dtype=tf.int32)
    return idx

key = random.PRNGKey(758493)
image_enc = jax.random.uniform(key, [batch_size, vis_seq_len, hidden_dim])
num_image_spans = vis_seq_len
num_image_patches_trg = int(num_image_spans * mask_rate)
image_patch_trg_idx = jnp.asarray(uniform_random_select(n=num_image_spans, num_samples=num_image_patches_trg, sort_idx=True)[None])

def produce_tensor_with_mask_token(image_enc, image_patch_trg_idx, num_image_spans):
    mask_one_hot = tf.reduce_any(tf.equal(tf.range(num_image_spans)[:, None], image_patch_trg_idx), 1)[None]
    '''returns a ragged jax array 'new_image_inc' of shape (1, 1080, variable size) where 
	if mask_one_hot[x][y] = False
        	the new_image_inc[x][y] value will be same as the image_enc[x][y] value 
        else
        	the new_image_inc[x][y] value will be MASK token'''


```
Sorry again, does this work?

**Reply by jakevdp** [reply]
I see, thanks. JAX doesn't support ragged arrays (every row must be the same size), so one possibility would be to fill the full row with the `MASK` token. Would that work for your usecase?

**Reply by jakevdp** [reply]
I feel bad this took so long – I keyed in on the wrong thing on your original question (ShapedArray) where I should have keyed in on "ragged", which JAX doesn't support.

**Reply by snat1505027** [reply]
Thanks for the reply. I had another question. If I have a DeviceArray, `X`, of shape `(945, 4, 768) -> (row, column, hidden_dim) ` and another DeviceArray, `IDX`, of shape `(945, 1) `where each entry in `IDX` is a random integer value between` 0 to 3` indicating one of the column in each row of `X`. I want to return a new DeviceArray where for row `p`, the hidden_dim in the column mentioned in `IDX[p]` will be replaced by `tf.zeros([768])`.

```
import jax

rkey = jax.random.PRNGKey(758493)
seq_mask = jax.random.randint(rkey, [945, 1], minval=0, maxval=4)
seq = jax.random.uniform(rkey, [945, 4, 768])

def produce_masked_seq(seq, seq_mask):
      ''' returns new_seq with shape (945, 4, 768) where in each row of new_seq there will be one column with tf.zeros([768]) and that column will be given in seq_mask '''

```

Is there any in-line code for that? Please let me know if you feel the code/description is ambiguous. Thanks!

**Reply by jakevdp** [reply]
If I understand correctly, I think you're looking for something like `out = jax.vmap(lambda x, i: x.at[i].set(0))(seq, seq_mask)`
